### PR TITLE
#8211: Map switcher and legend widgets

### PR DESCRIPTION
**Pull request by dsuren1**
## Description
This PR adds new feature to dashboard map and legend widgets
1. Map widget now lets user to select multiple maps and configure them with the support of Map switcher
2. Legend widget now allows user to change layer settings (visibility, opacity), legend looks similar to layers in TOC

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

## Issue

**What is the current behavior?**
#8211 

**What is the new behavior?**
https://docs.google.com/document/d/1cRGdMZGDhFGSBVjxAkSG__Vyty9FAhIA2grVserH9Xc/edit

legend-ex

wid-add-layer

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

## Other useful information

Reference: https://github.com/geosolutions-it/MapStore2-C040/issues/575

### CODE REVIEW

File: web/client/components/widgets/builder/wizard/map/MapCatalog.jsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+const SideGrid = compose(
+    loadingState(({ loading, items = [] }) => items.length === 0 && loading),
+    emptyState(
+        ({ loading, items = [] }) => items.length === 0 && !loading,
+        {
+            title: <Message msgId="catalog.noRecordsMatched" />,
+            style: { transform: "translateY(50%)" }
+        })
+
+)(require('../../../../misc/cardgrids/SideGrid').default);
```
Comment (MV88) [inline web/client/components/widgets/builder/wizard/map/MapCatalog.jsx:29]: please use import statement here

File: web/client/components/widgets/builder/wizard/map/MapCatalog.jsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+export default ({
```
Comment (MV88) [inline web/client/components/widgets/builder/wizard/map/MapCatalog.jsx:31]: please let's give a name default export as a general rule

**Review by MV88 — CHANGES_REQUESTED** [review]
@dsuren1 on the code point of view I did not spot big things so my focus was more on the functionality point of view

@tdipisa 
here is a list of things I noticed. probably some of them can be excluded or moved to a different issue

-  [ ] you could potentially select  all the maps you have in the list,
can this be a configurable number, or do we not care?

- [x] sometimes an horizontal scrollbar appears
image

- [x] name of widget is covered if no space is available
image

- [ ] it's not clear that you have to press CTRL to select multiple cards/maps
- [x] I have noticed some connection problems in particular when a  table widget is connect to a map and then if you change the map and then select a map that does not have the layer used by the table among layers

https://user-images.githubusercontent.com/11991428/171450373-dfdd1ce4-6f4d-484c-bc40-7193820d955f.mp4

- [x] do we have to show something if legend is not available?
image
in another case we have a message
image

- [x] is it normal to have legends continuously re-rendered?
https://user-images.githubusercontent.com/11991428/171452793-d0876411-d6c0-46a1-a5c2-e98a1d623458.mp4

**Comment by dsuren1** [comment]
@MV88 
Thanks.

> * [ ]  name of widget is covered if no space is available

_I'm thinking of showing an icon when widget width is small. As we expand the widget when the width is enough to fit map switcher we display the same. Your thoughts?_
image

> * [ ]  it's not clear that you have to press CTRL to select multiple cards/maps

_I thought since layers in TOC follows the same, it doesn't need any explicit helptext. Shall I add something like this?_
image

> * [ ]  I have noticed some connection problems in particular when a  table widget is connect to a map and then if you change the map and then select a map that does not have the layer used by the table among layers
> 
>  sync.connection.problem.mp4 

_I think this scenario is produced this way_
_1. Create map widget_
_2. Create table widget_
_3. Connect table with map_
_4. Delete map widget_ 
_5. Now add new map widget_
_6. Connection doesn't work. Because old connection is not cleared when dependant widget is deleted. Seems to be an existing issue. I will fix this_

> * [ ]  do we have to show something if legend is not available?

_This is due to the layer being `3dtiles`. I will filter the layers accordingly, such that it doesn't display any `3dtiles` layer in legend widget/wizard_

> * [ ]  is it normal to have legends continuously re-rendered?
>   https://user-images.githubusercontent.com/11991428/171452793-d0876411-d6c0-46a1-a5c2-e98a1d623458.mp4

_Possibly coz of [this]( https://github.com/geosolutions-it/MapStore2/blob/master/web/client/components/TOC/fragments/legend/Legend.jsx#L57) when `0` is used as `urlIdx`, and url is array it tends to generate a random number causing it to rerender. I will fix this_

**Comment by dsuren1** [comment]
> @dsuren1 thanks for commenting it out
> 
> > > * [ ]    name of widget is covered if no space is available
> > 
> > 
> > _I'm thinking of showing an icon when widget width is small. As we expand the widget when the width is enough to fit map switcher we display the same. Your thoughts?_ image
> 
> it could be a solution, unless you have already implemented we can consider it as a future enhancement

_I have implemented it along other updated in this PR_

> 
> > > * [ ]    it's not clear that you have to press CTRL to select multiple cards/maps
> > 
> > 
> > _I thought since layers in TOC follows the same, it doesn't need any explicit helptext. Shall I add something like this?_ image
> 
> You are right (I did not knew you could to that also in the TOC). Anyway here it is more obvious that a user can select multiple items in the list, so I would specify it, we should also consider that CTRL is not present in MAC OS, please amend it considering this mac os aspect. maybe it could be similar to the previous proposal to use a icon showing a tooltip to explain selection @allyoucanmap what do you think here?
> 
> @dsuren1 So apart from first consideration, have you covered all the rest, so i can re-check right?

_Sure. I can add a tooltip instead of sub text including mac os's command key (⌘)_

**Comment by MV88** [comment]
> Sure. I can add a tooltip instead of sub text including mac os's command key (⌘)

i would change it as the following:

"
for multi-selection 
on windows use CTRL + click
on mac use control key + click
"

**Review by MV88 — CHANGES_REQUESTED** [review]
- [x] how big should be a map widget to see the map switcher?

image

I think you can revert this change since we have little time

- [ ] there are still some cases where dependencies are not correctly updated
see this

https://user-images.githubusercontent.com/11991428/172162929-aeca3db0-4a04-4b2a-8073-624da008ae2d.mp4

**Comment by dsuren1** [comment]
> * [ ]  how big should be a map widget to see the map switcher?
> 
> image
> 
> I think you can revert this change since we have little time

_The width of the widget should be >800px so it can fit the title and switcher_

> 
> * [ ]  there are still some cases where dependencies are not correctly updated
>   see this
> 
>  sync.problem.mp4

_Let me check_

**Comment by MV88** [comment]
> The width of the widget should be >800px so it can fit the title and switcher

then I would make it smaller, like > 400 which includes when widgets is large at least 2 columns, or you revert it to what it was before

**Review by MV88 — APPROVED** [review]
LGTM

**Comment by DavidQuartz** [comment]
@tdipisa @dsuren1 the multiple map selection feature 'control + click' does not work on Mac. It triggers a 'right click' effect instead.

**Pull request merged**